lib/rust/flexer/src/group.rs

```diff
@@ -111,120 +225,131 @@ impl From<&Group> for NFA {
-    use crate::automata::dfa::DFA;
     use crate::automata::nfa;
-    use crate::automata::nfa::NFA;
     use crate::automata::pattern::Pattern;
-    use crate::group::Group;
+    use crate::group::{Group, GroupRegistry};
     use crate::group::rule::Rule;
 
     use std::default::Default;
     use test::Bencher;
+    use enso_prelude::default;
 
-    fn newline() -> Group {
+    fn newline() -> GroupRegistry {
         let     pattern = Pattern::char('\n');
         let mut group   = Group::default();
 
         group.add_rule(Rule{pattern,callback:"".into()});
 
-        group
+        let mut registry = GroupRegistry::default();
+        registry.add_group(group);
+        registry
     }
 
-    fn letter() -> Group {
+    fn letter() -> GroupRegistry {
         let     pattern = Pattern::range('a'..='z');
         let mut group   = Group::default();
 
         group.add_rule(Rule{pattern,callback:"".into()});
 
-        group
+        let mut registry = GroupRegistry::default();
+        registry.add_group(group);
+        registry
     }
 
-    fn spaces() -> Group {
+    fn spaces() -> GroupRegistry {
         let     pattern = Pattern::char(' ').many1();
         let mut group   = Group::default();
 
         group.add_rule(Rule{pattern,callback:"".into()});
 
-        group
+        let mut registry = GroupRegistry::default();
+        registry.add_group(group);
+        registry
     }
 
-    fn letter_and_spaces() -> Group {
+    fn letter_and_spaces() -> GroupRegistry {
         let     letter = Pattern::range('a'..='z');
         let     spaces = Pattern::char(' ').many1();
         let mut group  = Group::default();
 
         group.add_rule(Rule{pattern:letter,callback:"".into()});
         group.add_rule(Rule{pattern:spaces,callback:"".into()});
 
-        group
+        let mut registry = GroupRegistry::default();
+        registry.add_group(group);
+        registry
     }
 
-    fn complex_rules(count:usize) -> Group {
+    fn complex_rules(count:usize) -> GroupRegistry {
         let mut group   = Group::default();
 
         for ix in 0..count {
             let string  = ix.to_string();
-            let all     = Pattern::all(&string);
-            let any     = Pattern::any(&string);
-            let none    = Pattern::none(&string);
-            let pattern = Pattern::many(all & any & none);
+            let all     = Pattern::all_of(&string);
+            let any     = Pattern::any_of(&string);
+            let none    = Pattern::none_of(&string);
+            let pattern = Pattern::many(all >> any >> none);
             group.add_rule(Rule{pattern:pattern.clone(),callback:"".into()})
         }
-        group
+
+        let mut registry = GroupRegistry::default();
+        registry.add_group(group);
+        registry
     }
 
     #[test]
     fn test_to_nfa_newline() {
-        assert_eq!(NFA::from(&newline()), nfa::tests::newline());
+        assert_eq!(newline().to_nfa_from(default()).unwrap(),nfa::tests::newline());
```

Comment:
you are using `unwrap` here. Why? Wouldnt it be better to just compare it to `Some(...)` ? Its nicer when tests tell you that the output is incorrect than panic